packages/airnode-node/src/evm/requests/blocking.ts

```diff
@@ -1,51 +1,112 @@
-
+  const withdrawalsBySponsorAddress = keyBy(pendingWithdrawals, 'sponsorAddress');
   const initialState: ApiCallsWithLogs = { logs: [], apiCalls: [] };
 
-  const { logs, apiCalls } = pendingApiCalls.reduce((acc, apiCall) => {
+  const { logs, apiCalls } = requests.apiCalls.reduce((acc, apiCall) => {
     const pendingWithdrawal = withdrawalsBySponsorAddress[apiCall.sponsorAddress];
-
-    if (pendingWithdrawal) {
-      const warningLog = logger.pend(
-        'WARN',
-        `Ignoring Request ID:${apiCall.id} as it has a pending Withdrawl ID:${pendingWithdrawal.id}`
-      );
-      const blockedCall: Request<ApiCall> = {
-        ...apiCall,
-        status: RequestStatus.Ignored,
-        errorMessage: `${RequestErrorMessage.PendingWithdrawal}: ${pendingWithdrawal.id}`,
-      };
-      return {
-        ...acc,
-        logs: [...acc.logs, warningLog],
-        apiCalls: [...acc.apiCalls, blockedCall],
-      };
+    if (apiCall.status !== RequestStatus.Pending || !pendingWithdrawal) {
+      return { ...acc, apiCalls: [...acc.apiCalls, apiCall] };
     }
 
-    return { ...acc, apiCalls: [...acc.apiCalls, apiCall] };
+    const warningLog = logger.pend(
+      'WARN',
+      `Ignoring Request ID:${apiCall.id} as it has a pending Withdrawal ID:${pendingWithdrawal.id}`
+    );
+    const blockedCall: Request<ApiCall> = {
+      ...apiCall,
+      status: RequestStatus.Ignored,
+      // TODO: These error messages should be functions accepting id parameter and creating the error string
+      errorMessage: `${RequestErrorMessage.PendingWithdrawal}: ${pendingWithdrawal.id}`,
+    };
+
+    return {
+      ...acc,
+      logs: [...acc.logs, warningLog],
+      apiCalls: [...acc.apiCalls, blockedCall],
+    };
   }, initialState);
 
-  const updatedRequests = { ...requests, apiCalls };
-  return [logs, updatedRequests];
+  return [[...prevLogs, ...logs], { ...requests, apiCalls }];
+}
+
+/**
+ * We need to limit the requests from the same sponsor because they are performed in parallel. However, we cannot just
+ * filter based on sponsor or sponsor wallet only, because we have not yet validated which is the correct sponsor wallet
+ * for a given sponsor.
+ *
+ * Instead we need to limit the request based on (sponsor, sponsor wallet) pair.
+ */
+export function applySponsorAndSponsorWalletRequestLimit([
+  prevLogs,
+  requests,
+]: LogsData<GroupedRequests>): LogsData<GroupedRequests> {
+  const apiCallsWithSponsorId = requests.apiCalls.map((apiCall) => ({
+    ...apiCall,
+    // The "sponsor id" is a unique combination of sponsor and sponsor wallet
+    sponsorId: `${apiCall.sponsorAddress}-${apiCall.sponsorWalletAddress}`,
+  }));
+  const requestCountPerSponsorId = new Map<string, number>();
+  const allowedApiCalls: Request<ApiCall>[] = [];
+  const logs: PendingLog[] = [];
+
+  // TODO: Consider reduce or map for this as a functional approach
```

Comment:
Not really necessary

Reply:
Yeah, I thought so as well, but wanted feedback.